File: srcs/requirements/Pong-Server/server/game/Player.py
from config import *
from Vec2 import *
from Hitbox import *

class Player:
	def __init__(self, nb, name, nb_total, borderless, square):
		self.borderless = borderless
		self.nb_total = nb_total
		self.square = square
		self.speed_per_sec = speed_per_sec
		self.speed = self.speed_per_sec * 0.005
		self.nb = nb
		self.name = name
		self.win = "LOSE"
		self.score = 0
		self.initPaddle(nb_total, borderless, square)
		
  
	def initPaddle(self, nb_total, borderless, square):
		self.size = [winWidth * 0.007 ,winHeight * 0.1]
		if nb_total == 2:
			if (self.nb == 1):
				pos = [winWidth * 0.02, (winHeight / 2) - (self.size[1] / 2)]
				self.goal = Hitbox(-50, 0, 50, winHeight)
				self.side = "left"
			else:
				pos = [winWidth - (winWidth * 0.02) - self.size[0], (winHeight / 2) - (self.size[1] / 2)]
				self.goal = Hitbox(winWidth, 0, 50, winHeight)
				self.side = "right"

		elif square:
			if (self.nb == 1):
				pos = [winWidth * 0.02, (winHeight / 2) - (self.size[1] / 2)]
				self.goal = Hitbox(-43, 0, 50, winHeight)
				self.side = "left"
			elif (self.nb == 2):
				pos = [winWidth - (winWidth * 0.02) - self.size[0], (winHeight / 2) - (self.size[1] / 2)]
				self.goal = Hitbox(winWidth - 7, 0, 50, winHeight)
				self.side = "right"
			elif (self.nb == 3):
				self.size.reverse()
				self.size[0] = winWidth * 0.1
				self.speed_per_sec = winWidth
				pos = [(winWidth / 2) - (self.size[0] / 2), (winWidth * 0.02)]
				self.goal = Hitbox(0, -43, winWidth, 50)
				self.side = "up"
			else:
				self.size.reverse()
				self.size[0] = winWidth * 0.1
				self.speed_per_sec = winWidth
				pos = [(winWidth / 2) - (self.size[0] / 2), winHeight - (winWidth * 0.02) - self.size[1]]
				self.goal = Hitbox(0, winHeight - 7, winWidth, 50)
				self.side = "down"

		elif nb_total == 4:
			if (self.nb == 1):
				pos = [winWidth * 0.02, (winHeight / 2) - self.size[1]]
				self.goal = Hitbox(-50, 0, 50, winHeight)
				self.side = "left"
			elif (self.nb == 2):
				pos = [winWidth * 0.02 + 50, (winHeight / 2)]
				self.goal = Hitbox(-50, 0, 50, winHeight)
				self.side = "left"
			elif (self.nb == 3):
				pos = [winWidth - (winWidth * 0.02) - self.size[0], (winHeight / 2) - self.size[1]]
				self.goal = Hitbox(winWidth, 0, 50, winHeight)
				self.side = "right"
			else:
				pos = [winWidth - (winWidth * 0.02) - self.size[0] - 50, (winHeight / 2)]
				self.goal = Hitbox(winWidth, 0, 50, winHeight)
				self.side = "right"
     
		self.paddle = [Hitbox(pos=Vec2(pos=pos), size=self.size)]
		if borderless:
			self.paddle.append(Hitbox(pos[0], pos[1] - winHeight, size=self.size))
			self.paddle.append(Hitbox(pos[0], pos[1] + winHeight, size=self.size))
 
	def	moveUp(self, walls):
		init_speed = self.speed / 4
		tmp_speed = init_speed
		collision = False
		while walls and not collision and tmp_speed <= self.speed:
			tmp_pos = Vec2(self.paddle[0].pos.x, self.paddle[0].pos.y - tmp_speed)
			collision = is_colliding(tmp_pos, self.size, walls[0].hitbox.pos, walls[0].size)
			if collision:
				break
			tmp_speed += init_speed
		
		if not collision:
			self.paddle[0].pos.y -= self.speed
		else:
			self.paddle[0].pos.y = round(self.paddle[0].pos.y - (tmp_speed - init_speed))
			while not is_colliding(self.paddle[0].pos, self.size, walls[0].hitbox.pos, walls[0].size):
				self.paddle[0].pos.y -= 1

		if self.borderless:
			if self.paddle[0].pos.y + (self.size[1] / 2) < 0:
				self.paddle[0].pos.y += winHeight
			self.paddle[1].pos.y = self.paddle[0].pos.y - winHeight
			self.paddle[2].pos.y = self.paddle[0].pos.y + winHeight

			
	def moveDown(self, walls):
		init_speed = self.speed / 4
		tmp_speed = init_speed
		collision = False
		while walls and not collision and tmp_speed <= self.speed:
			tmp_pos = Vec2(self.paddle[0].pos.x, self.paddle[0].pos.y + tmp_speed)
			collision = is_colliding(tmp_pos, self.size, walls[1].hitbox.pos, walls[1].size)
			if collision:
				break
			tmp_speed += init_speed
		
		if not collision:
			self.paddle[0].pos.y += self.speed
		else:
			self.paddle[0].pos.y = round(self.paddle[0].pos.y + (tmp_speed - init_speed))
			while not is_colliding(self.paddle[0].pos, self.size, walls[1].hitbox.pos, walls[1].size):
				self.paddle[0].pos.y += 1
  
		if self.borderless:
			if self.paddle[0].pos.y + (self.size[1] / 2) >= winHeight:
				self.paddle[0].pos.y -= winHeight
			self.paddle[1].pos.y = self.paddle[0].pos.y - winHeight
			self.paddle[2].pos.y = self.paddle[0].pos.y + winHeight

	def moveLeft(self, walls):
		init_speed = self.speed / 4
		tmp_speed = init_speed
		collision = False
		while walls and not collision and tmp_speed <= self.speed:
			tmp_pos = Vec2(self.paddle[0].pos.x - tmp_speed, self.paddle[0].pos.y)
			collision = is_colliding(tmp_pos, self.size, walls[2].hitbox.pos, walls[2].size)
			if collision:
				break
			tmp_speed += init_speed
		
		if not collision:
			self.paddle[0].pos.x -= self.speed
		else:
			self.paddle[0].pos.x = round(self.paddle[0].pos.x - (tmp_speed - init_speed))
			while not is_colliding(self.paddle[0].pos, self.size, walls[2].hitbox.pos, walls[2].size):
				self.paddle[0].pos.x -= 1
   
	def moveRight(self, walls):
		init_speed = self.speed / 4
		tmp_speed = init_speed
		collision = False
		while walls and not collision and tmp_speed <= self.speed:
			tmp_pos = Vec2(self.paddle[0].pos.x + tmp_speed, self.paddle[0].pos.y)
			collision = is_colliding(tmp_pos, self.size, walls[3].hitbox.pos, walls[3].size)
			if collision:
				break
			tmp_speed += init_speed
		
		if not collision:
			self.paddle[0].pos.x += self.speed
		else:
			self.paddle[0].pos.x = round(self.paddle[0].pos.x + (tmp_speed - init_speed))
			while not is_colliding(self.paddle[0].pos, self.size, walls[3].hitbox.pos, walls[3].size):
				self.paddle[0].pos.x += 1


	def collide(self, ball):
		for paddle in self.paddle:
			ball_box = [ball.center[0].x - ball.radius, ball.center[0].y - ball.radius]
			if not is_colliding(ball_box, [ball.radius * 2, ball.radius * 2], paddle.pos, self.size):
				continue
			paddle.collidePaddle(ball, self.nb)
			break

	def update(self, delta):
		if self.borderless:
			self.paddle[1].pos.y = self.paddle[0].pos.y - winHeight
			self.paddle[2].pos.y = self.paddle[0].pos.y + winHeight
		self.speed = self.speed_per_sec * delta

File: srcs/django_code/front/js/game/Player.js
import { Vec2 } from "./Vec2.js";
import { Hitbox, is_colliding } from "./Hitbox.js";
import { canvas, ctx } from "./canvas.js";

export class Player {
	constructor(nb, name, nb_total, borderless, square) {
		this.speed_per_sec = canvas.height;
		this.speed = this.speed_per_sec * 0.01;
		this.nb = nb;
		this.name = name;
		this.square = square;
		this.borderless = borderless;
		this.win = "LOSE";
		this.score = 0;
		this.initPaddle(nb_total, borderless, square);
	}

	initPaddle(nb_total, borderless, square) {
		this.size = [canvas.width * 0.007, canvas.height * 0.1];
		let pos;
		if (nb_total == 2) {
			if (this.nb == 1) {
				pos = [canvas.width * 0.02, (canvas.height / 2) - (this.size[1] / 2)];
				this.goal = new Hitbox(-50, 0, 50, canvas.height);
				this.side = "left";
			}
			else {
				pos = [canvas.width * 0.98 - this.size[0], (canvas.height / 2) - (this.size[1] / 2)];
				this.goal = new Hitbox(canvas.width, 0, 50, canvas.height);
				this.side = "right";
			}
		}
		else if (square) {
			const goal_size = canvas.height * 0.0125;
			if (this.nb == 1) {
				pos = [canvas.width * 0.02, (canvas.height / 2) - (this.size[1] / 2)];
				this.goal = new Hitbox(-50 + goal_size, 0, 50, canvas.height);
				this.side = "left";
			}
			else if (this.nb == 2) {
				pos = [canvas.width * 0.98 - this.size[0], (canvas.height / 2) - (this.size[1] / 2)];
				this.goal = new Hitbox(canvas.width - goal_size, 0, 50, canvas.height);
				this.side = "right";
			}
			else if (this.nb == 3) {
				this.size.reverse();
				this.size[0] = canvas.width * 0.1;
				this.speed_per_sec = canvas.width;
				pos = [(canvas.width / 2) - (this.size[0] / 2), canvas.height * 0.02];
				this.goal = new Hitbox(0, -50 + goal_size, canvas.width, 50);
				this.side = "up";
			}
			else {
				this.size.reverse();
				this.size[0] = canvas.width * 0.1;
				this.speed_per_sec = canvas.width;
				pos = [(canvas.width / 2) - (this.size[0] / 2), canvas.height * 0.98 - this.size[1]];
				this.goal = new Hitbox(0, canvas.height - goal_size, canvas.width, 50);
				this.side = "down";
			}
		}
		else {
			if (this.nb == 1) {
				pos = [canvas.width * 0.02, (canvas.height / 2) - this.size[1]];
				this.goal = new Hitbox(-50, 0, 50, canvas.height);
				this.side = "left";
			}
			else if (this.nb == 2){
				pos = [canvas.width * 0.02 + 50, (canvas.height / 2)];
				this.goal = new Hitbox(-50, 0, 50, canvas.height);
				this.side = "left";
			}
			else if (this.nb == 3) {
				pos = [canvas.width * 0.98 - this.size[0], (canvas.height / 2) - this.size[1]];
				this.goal = new Hitbox(canvas.width, 0, 50, canvas.height);
				this.side = "right";
			}
			else {
				pos = [canvas.width * 0.98 - this.size[0] - 50, (canvas.height / 2)];
				this.goal = new Hitbox(canvas.width, 0, 50, canvas.height);
				this.side = "right";
			}
		}
		this.paddle = [new Hitbox(pos[0], pos[1], this.size[0], this.size[1])];
		if (borderless) {
			this.paddle.push(new Hitbox(pos[0], pos[1] - canvas.height, this.size[0], this.size[1]));
			this.paddle.push(new Hitbox(pos[0], pos[1] + canvas.height, this.size[0], this.size[1]));
		}
	}

	moveUp(walls) {
		const init_speed = this.speed / 4;
		let tmp_speed = init_speed;
		let collision = false;
		let tmp_pos;
		while (walls && !collision && tmp_speed <= this.speed) {
			tmp_pos = new Vec2(this.paddle[0].pos.x, this.paddle[0].pos.y - tmp_speed);
			collision = is_colliding(tmp_pos, this.size, walls[0].Hitbox.pos, walls[0].size);
			if (collision)
				break;
			tmp_speed += init_speed;
		}
		if (!collision){
			this.paddle[0].pos.y -= this.speed;
		}
		else {
			this.paddle[0].pos.y = Math.round(this.paddle[0].pos.y - (tmp_speed - init_speed));
			while (!is_colliding(this.paddle[0].pos, this.size, walls[0].Hitbox.pos, walls[0].size))
				this.paddle[0].pos.y -= 1;
		}

		if (this.borderless) 
			if (this.paddle[0].pos.y + (this.size[1] / 2) < 0)
				this.paddle[0].pos.y += canvas.height;
	}

	moveDown(walls) {
		const init_speed = this.speed / 4;
		let tmp_speed = init_speed;
		let collision = false;
		let tmp_pos;
		while (walls && !collision && tmp_speed <= this.speed) {
			tmp_pos = new Vec2(this.paddle[0].pos.x, this.paddle[0].pos.y + tmp_speed);
			collision = is_colliding(tmp_pos, this.size, walls[1].Hitbox.pos, walls[1].size);
			if (collision)
				break;
			tmp_speed += init_speed;
		}
		if (!collision)
			this.paddle[0].pos.y += this.speed;
		else {
			this.paddle[0].pos.y = Math.round(this.paddle[0].pos.y + (tmp_speed - init_speed));
			while (!is_colliding(this.paddle[0].pos, this.size, walls[1].Hitbox.pos, walls[1].size))
				this.paddle[0].pos.y += 1;
		}

		if (this.borderless)
			if (this.paddle[0].pos.y + (this.size[1] / 2) > canvas.height)
				this.paddle[0].pos.y -= canvas.height;
	}

	moveLeft(walls) {
		const init_speed = this.speed / 4;
		let tmp_speed = init_speed;
		let collision = false;
		let tmp_pos;
		while (walls && !collision && tmp_speed <= this.speed) {
			tmp_pos = new Vec2(this.paddle[0].pos.x - tmp_speed, this.paddle[0].pos.y);
			collision = is_colliding(tmp_pos, this.size, walls[2].Hitbox.pos, walls[2].size);
			if (collision)
				break;
			tmp_speed += init_speed;
		}
		if (!collision)
			this.paddle[0].pos.x -= this.speed;
		else {
			this.paddle[0].pos.x = Math.round(this.paddle[0].pos.x - (tmp_speed - init_speed));
			while (!is_colliding(this.paddle[0].pos, this.size, walls[2].Hitbox.pos, walls[2].size))
				this.paddle[0].pos.x -= 1;
		}
	}

	moveRight(walls) {
		const init_speed = this.speed / 4;
		let tmp_speed = init_speed;
		let collision = false;
		let tmp_pos;
		while (walls && !collision && tmp_speed <= this.speed) {
			tmp_pos = new Vec2(this.paddle[0].pos.x + tmp_speed, this.paddle[0].pos.y);
			collision = is_colliding(tmp_pos, this.size, walls[3].Hitbox.pos, walls[3].size);
			if (collision)
				break;
			tmp_speed += init_speed;
		}
		if (!collision)
			this.paddle[0].pos.x += this.speed;
		else {
			this.paddle[0].pos.x = Math.round(this.paddle[0].pos.x + (tmp_speed - init_speed));
			while (!is_colliding(this.paddle[0].pos, this.size, walls[3].Hitbox.pos, walls[3].size))
				this.paddle[0].pos.x += 1;
		}
	}

	draw() {
		for (let paddle of this.paddle)
			ctx.fillRect(paddle.pos.x, paddle.pos.y, this.size[0], this.size[1]);
	}

	collide(ball) {
		const ball_box = new Vec2(ball.center[0].x - ball.radius, ball.center[0].y - ball.radius);
		for (let paddle of this.paddle) {
			if (!is_colliding(ball_box, [ball.radius * 2, ball.radius * 2], paddle.pos, this.size))
				continue;
			paddle.collidePaddle(ball, this.nb);
			break;
		}
	}

	update(delta) {
		if (this.borderless) {
			this.paddle[1].pos.y = this.paddle[0].pos.y - canvas.height;
			this.paddle[2].pos.y = this.paddle[0].pos.y + canvas.height;
		}
		this.speed = this.speed_per_sec * delta;
	}

	responsive(old_sizes) {
		this.speed_per_sec = canvas.height;
		this.size = [canvas.width * 0.007, canvas.height * 0.1];
		if (this.side == "up" || this.side == "down") {
			this.size.reverse();
			this.size[0] = canvas.width * 0.1;
		}
		if (this.square) {
			const goal_size = canvas.height * 0.0125;
			if (this.side == "left")
				this.goal = new Hitbox(-50 + goal_size, 0, 50, canvas.height);
			else if (this.side == "right")
				this.goal = new Hitbox(canvas.width - goal_size, 0, 50, canvas.height);
			else if (this.side == "up")
				this.goal = new Hitbox(0, -50 + goal_size, canvas.width, 50);
			else
				this.goal = new Hitbox(0, canvas.height - goal_size, canvas.width, 50);
		}
		else {
			if (this.side == "left")
				this.goal = new Hitbox(-50, 0, 50, canvas.height);
			else
				this.goal = new Hitbox(canvas.width, 0, 50, canvas.height);
		}

		const pos_ratio = [this.paddle[0].pos.x / old_sizes[0], this.paddle[0].pos.y / old_sizes[1]];
		this.paddle[0] = new Hitbox(pos_ratio[0] * canvas.width, pos_ratio[1] * canvas.height, this.size[0], this.size[1]);
		if (this.borderless) {
			this.paddle[1] = new Hitbox(this.paddle[0].pos.x, this.paddle[0].pos.y - canvas.height, this.size[0], this.size[1]);
			this.paddle[2] = new Hitbox(this.paddle[0].pos.x, this.paddle[0].pos.y + canvas.height, this.size[0], this.size[1]);
		}
	}
}